text/0008-ingest-gem-content.md

```diff
@@ -0,0 +1,332 @@
+The saved output will include file size, content type, SHA256 checksum, and relative path within the gem.
+In the interest of simplifying this RFC and the feature work that follows, this RFC focuses on only the ingestion and storage of gem content.
+
+# Motivation
+
+It is our hope that we can provide access to individual file data from gem files across versions in a compact and efficient way that supports future features depending on gem content.
+
+Creating a gem content ingestion and storage system will enable rubygems.org to support the security and useability gems in the Ruby Gems ecosystem.
+The gem content repository will be the authoritative source for the content of gems, accessible as individual files without without requiring downloading and unpacking a gem.
+
+In the near future, the indexed gem content will allow an API for accessing the gem content.
+The API will support both rubygems.org and external projects that wish to use individual gem content.
+Rubygems.org plans to add tools for browsing gem content on rubygems.org or through the `gem` command line interface.
+
+# Guide-level explanation
+
+For the purpose of illustration, I will use the gem `rake` as an example.
+
+When a gem is uploaded, rubygems.org saves it as a tarball file named `rake-13.0.6.gem`.
+A gem may be installed with `gem install` or the `.gem` file can be downloaded using `gem fetch`.
+Downloading the entire gem locally is currently the only way to access the contents of gem.
+
+With the gem content ingestion system in place, gem contents will be saved to S3 and indexed in the rubygems.org database.
+Binary files will be recorded in metadata, but not stored.
+Gem contents will be stored at an S3 key named for the SHA256 of the file to de-duplicate files within a gem.
+Yanking will remove file contents unique to the gem, as expected, without altering unyanked gem contents.
+
+### Background process
+
+In order to ingest the contents of a gem, a background process can be enqueued to index and store the gem contents.
+
+At gem upload time, or on demand, a background worker will download, unpack, record file metadata, index files in the database as rows, and upload each file to a content-only S3 bucket.
+The process will result in a row in the database for each file in a gem, with the relative path, checksum, and size of every file.
+
+### Indexing files to create a manifest
+
+The background worker will create a json document manifest of all the files and store it in S3 alongside the files themselves.
+
+The manifest will hold the path, checksum, size, content type, binary determination, and plain text line count.
+
+In order to support yanking gems, which requires comparing checksums across all versions of a gem, a table will be added to the rubygems.org database.
+The table will consist of one row per gem version with a foreign key for the `version_id` and a jsonb column holding only the SHA256 checksums of all files in that version.
```

Comment:
why one row per version, and not per version per file? I think if each version/file had its own row, it would make garbage collection much easier, since for each file as we delete it, we can check to see if that was the only reference with the same checksum

Reply:
I’m worried that one row per version per version per file will produce a table with hundreds of millions of rows and potentially congest the primary database—though I’m open to other suggestions about how to avoid that.

Reply:
Yes, this was a compromise to try to reduce the number of rows. The rows will never be used individually, but only as a group for each gem version. Some rough pseudo code to yank a version would look like:
```
keep_files = load_all_checksums_excluding(yanked_version)
yanked_files = checksums_for(yanked_version)
remove_files = yanked_files - keep_files # remove any files that exist in another version.
storage.delete_files(remove_files)
delete_version_checksums(yanked_version)
```
This is the only interaction where we would use this table.

Reply:
hmmmm....

I was thinking we'd use the rows to enable redirects:

`/gems/rake/versions/13.0.0/content/lib/rake.rb` -> `s3 backed domain/gems/rake/$HASH` ?

Reply:
That redirect pattern is one of the benefits.

On the contrary, unless you're already viewing the source somewhere else, knowing that `lib/rake.rb` exists at all means you already requested the manifest and you therefore have all the data for mapping path to hash.

Reply:
The more I think about this the more I am leaning towards a second RDS Postgres that hold just the table of one-row-per-file-per-version? It's somewhat higher infra complexity to have a second postgres, but I think it's much lower codebase and operational complexity because the contents API loses the ability to take down any main site functionality.

That way we can index on both path and content hash, get single row results immediately, and never worry about congesting the main database. Rails 6/7 has excellent support for models that talk to a different database, and if it's in a separate postgres instance I would feel completely comfortable with a single table with hundreds of millions of rows.

Reply:
@indirect Is the purpose of this to be able to load a file by a nice path rather than a hash? We could arguably tell S3 to configure this bucket for static web hosting, and then add [static redirects](https://docs.aws.amazon.com/AmazonS3/latest/userguide/how-to-page-redirect.html) for all file paths to the appropriate hash.

Reply:
Between json blob and rows with only version and sha, I think the size will be about the same, but if you add the necessity of an index on the global hashing, one row per file, you will add that large index that is not needed in the per gem approach.

If you want path and want to perform redirects, then I would not use the json blob approach. This would also change this table to be significantly more read heavy than write heavy.

Reply:
I think ~100,000,000 static redirects for a single S3 bucket is... probably too many? I haven't checked on the limits, though.

Reply:
Personally I don't think the pretty urls are worth the extra complexity since almost every request ever will use the eventual website user interface which _must_ load the manifest first and will automatically hide any client side actions that load data. I never imagine sending anyone to the actual raw url we're discussing here with the path in it.